### PR TITLE
[RISCV] Add Qualcomm uC Xqcisls (Scaled Load Store) extension

### PR DESCRIPTION
**Pull request by svs-quic**
This extension adds 8 load/store instructions with a scaled index addressing mode.

The current spec can be found at:
https://github.com/quic/riscv-unified-db/releases/latest

This patch adds assembler only support.

### CODE REVIEW

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 1]
```diff
@@ -771,9 +772,11 @@ Error RISCVISAInfo::checkDependency() {
...
   }
 
-  if (Exts.count("xqcicsr") != 0 && (XLen != 32)) {
-    return getError("'xqcicsr' is only supported for 'rv32'");
-  }
+  for (auto Ext : XqciExts)
+    if (Exts.count(Ext.str()) && (XLen != 32))
+      return createStringError(errc::invalid_argument,
+                               "'" + Twine(Ext) + "'" +
+                                   " is only supported for 'rv32'");
```
Comment (lenary) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:779]: There aren't any tests for this error - please may you add some in `RISCVISAInfoTest.cpp`?

File: llvm/lib/Target/RISCV/RISCVInstrInfoXqci.td [hunk 2]
```diff
@@ -22,6 +22,28 @@
...
+  let Inst{31-28} = func4;
+  let Inst{27-25} = shamt;
+}
+}
+
+let hasSideEffects = 0, mayLoad = 0, mayStore = 1 in {
+class QCIStore_ScaleIdx<bits<4> func4, string opcodestr>
+    : RVInstRBase<0b110, OPC_CUSTOM_1, (outs),
+                  (ins GPR:$rd, GPRMem:$rs1, GPRNoX0:$rs2, uimm3:$shamt),
+                  opcodestr, "$rd, $rs1, $rs2, $shamt"> {
```
Comment (lenary) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXqci.td:40]: Can you add a comment to this class, noting that `rd` corresponds to the field in the PDF description named `rs3` (as it's a source for the stores, not a destination)

**Review by lenary — COMMENTED** [review]
The encodings and functionality here are correct, but we're missing a test and it would be good to add a comment to the scaled store tablegen class.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp (hunk 1 above)
Comment (svs-quic) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:779]: Thanks Sam. I've added a test in RISCVISAInfoTest.cpp

**Review by lenary — APPROVED** [review]
LGTM

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 3]
```diff
@@ -771,9 +772,11 @@ Error RISCVISAInfo::checkDependency() {
       return getIncompatibleError("xwchc", "zcb");
   }
 
-  if (Exts.count("xqcicsr") != 0 && (XLen != 32)) {
-    return getError("'xqcicsr' is only supported for 'rv32'");
-  }
+  for (auto Ext : XqciExts)
+    if (Exts.count(Ext.str()) && (XLen != 32))
+      return createStringError(errc::invalid_argument,
```
[the remark this answers is not in the document]
Comment (svs-quic) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp]: Thanks, I've made the change.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**